src/proto/hybrid_vec.rs

```diff
@@ -0,0 +1,1832 @@
+
+    fn next(&mut self) -> Option<T> {
+        match *self {
+            IntoIter::Stack { ref array, ref mut curr, len } if *curr != len => { 
+                let ret = unsafe { ptr::read(&array[*curr]) };
+                *curr += 1;
+                Some(ret)
+            },
+            IntoIter::Heap { ptr: _, cap: _, ref mut curr, end } if *curr != end => unsafe {
+                if mem::size_of::<T>() == 0 {
+                    *curr = mem::transmute(*curr as usize + 1);
+                    return Some(ptr::read(mem::transmute(1us)));
+                }
+
+                let ret = ptr::read(*curr);
+                *curr = curr.offset(1);
+                Some(ret)
+            },
+            _ => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let exact = match *self {
+            IntoIter::Stack { array: _, curr, len } => len - curr,
+            IntoIter::Heap { ptr: _, cap: _, curr, end } => {
+                let diff = (end as usize) - (curr as usize);
+                let size = mem::size_of::<T>();
+                diff / (if size == 0 {1} else {size})
+            }
+        };
+        
+        (exact, Some(exact))  
+    }
+}
+
+impl<T> DoubleEndedIterator for IntoIter<T> {
+    //#[inline]
+    fn next_back<'a>(&'a mut self) -> Option<T> {
+        match *self {
+            IntoIter::Stack { ref mut array, curr, ref mut len } if curr != *len => {
+                *len -= 1;
+                Some(unsafe { ptr::read(&array[*len]) })
+            },
+            IntoIter::Heap { ptr: _, cap: _, curr, ref mut end } if curr != *end => unsafe {
+                if mem::size_of::<T>() == 0 {
+                    *end = mem::transmute(*end as usize - 1);
+                    Some(ptr::read(mem::transmute(1us)))
+                } else {
+                    *end = end.offset(-1);
+                    Some(ptr::read(mem::transmute(*end)))
+                }
+            },
+            _ => None,
+        }
+    }
+}
+
+impl<T> ExactSizeIterator for IntoIter<T> {}
+
+impl<T> IntoIter<T> { 
+    pub fn into_inner(mut self) -> HybridVec<T> {
+        for _ in self {}
+
+        let ret = match self {
+            IntoIter::Stack { ref array, curr: _, len: _ } =>
+            HybridVec::Stack {
+                array: unsafe { ptr::read(array) },
+                len: 0,
+            },
+            IntoIter::Heap { ptr, cap, curr: _, end: _, } => unsafe { 
+                HybridVec::from_raw_parts(ptr, cap, 0) 
+            },
+        };
+
+        unsafe { mem::forget(self); }
+        ret
+    }    
+}
+
+#[unsafe_destructor]
+impl<T> Drop for IntoIter<T> {
+    fn drop(&mut self) {
+        for _ in *self {}
+
+        unsafe { 
+            let temp = mem::replace(self, mem::zeroed());
+            match temp {
+                IntoIter::Heap { ptr, cap, curr: _, end: _ } if cap != 0 => dealloc(ptr, cap),
+                _ => (),
+            }
+
+            mem::forget(temp);
```

Comment:
@huonw @tbu- I am using `mem::replace` to swap `self` into a stack-local temporary and replace it with the empty heap variant. I edited this line:

```rust
let temp = mem::replace(self, HybridVec::Heap { ptr: nonzero_ptr(), len: 0, cap: 0 })
```

This segfaults in the drop glue:

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffff6ecd700 (LWP 32086)]
0x00005555556f9d35 in collect::proto::hybrid_vec::tests::DropCounter<'a>.Drop::drop (self=0x7ffff6ecbdf0) at hybrid_vec.rs:931
931	            *self.count += 1;
(gdb) bt
#0  0x00005555556f9d35 in collect::proto::hybrid_vec::tests::DropCounter<'a>.Drop::drop (self=0x7ffff6ecbdf0) at hybrid_vec.rs:931
#1  0x00005555556fc416 in proto..hybrid_vec..tests..DropCounter::glue_drop.25904::h58a7d1eb6e1fe9f1 ()
#2  0x00005555556fc548 in _$u{5b}proto..hybrid_vec..tests..DropCounter$u{3b}$u{20}32$u{5d}::glue_drop.25907::hc81ee838e92e2e29 ()
#3  0x00005555556fb31c in proto..hybrid_vec..HybridVec$LT$proto..hybrid_vec..tests..DropCounter$GT$::glue_drop.25901::hf8294716f328cfc8 ()
#4  0x00005555556fd8c2 in mem::drop::h11352505897820767777 ()
#5  0x00005555556fac7b in collect::proto::hybrid_vec::tests::test_double_drop () at hybrid_vec.rs:959
#6  0x0000555555783a91 in thunk::F.Invoke$LT$A$C$$u{20}R$GT$::invoke::h8737199544904279263 ()
#7  0x000055555578d79c in thunk::F.Invoke$LT$A$C$$u{20}R$GT$::invoke::h12075299542435467671 ()
#8  0x000055555578494f in rt::unwind::try::try_fn::h4864347942857532199 ()
#9  0x00005555557d6f59 in rust_try_inner ()
#10 0x00005555557d6f46 in rust_try ()
#11 0x0000555555784fa0 in thunk::F.Invoke$LT$A$C$$u{20}R$GT$::invoke::h9718536689904751371 ()
#12 0x00005555557d1586 in sys::thread::thread_start::h3f192e5a493c4515E3w ()
#13 0x00007ffff79c0182 in start_thread (arg=0x7ffff6ecd700) at pthread_create.c:312
#14 0x00007ffff72cf00d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:111
```